docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+* Failures among the workers can be recovered easily by restarting the failed workers and then load the current model sent from
+any of the existing workers. Model does not need to be replicated to support fault-tolerance, which avoids the waste of resources.
+* The model can be updated more efficiently by fully leveraging the network structure and collective communication algorithms.
+For example, in [ring-allreduce](http://research.baidu.com/bringing-hpc-techniques-deep-learning/) algorithm, each of *N* workers only needs to communicate with two of its peer workers *2 * (N − 1)* times
+to update all the model parameters completely.
+* Scaling up and down of the number of workers is as easy as reconstructing the underlying allreduce communicator and
+re-assigning the ranks among the workers.
+
+We'd like to support allreduce-based distributed training in ElasticDL so our users may obtain more efficient distributed
+training performance in certain use cases without having to worry about many challenges involved in performance tuning and
+resource allocation under parameter-server-based distributed training.
+
+In the following sections, we will explain the design of allreduce-based distributed training in ElasticDL in detail.
+
+For details on the existing technologies relevant to collective communications, please head over to the last section of
+this design doc.
+
+## Design Components
+
+There are two major design components: the fault-tolerant allreduce implementation and support for allreduce-based training in ElasticDL.
+The next two sections will illustrate the details of these components.
+
+### Fault-tolerant Allreduce Implementation
+
+We are collaborating with [Caicloud](https://github.com/caicloud/) on building an API that provides implementations of
+fault-tolerant allreduce. The initial implementation will contain an experimental Python binding for NCCL that is
+fault-tolerant and Kubernetes-native. This will include but not limited to the following objectives (more details to be disclosed later once
+the implementation has been open-sourced):
+
+* Fault-tolerant: if any of the worker pod fails, the [NCCL communicator](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html)
+can be reconstructed. The allreduce operation continues as long as there's at least one healthy worker pod.
+* Elastic: the number of worker pods can be dynamically added if there are enough computational resources available.
+The NCCL communicator will be reconstructed and the ranks can be re-assigned as the number of worker pods changes.
+
+The interface for allreduce would look like the following:
+
+```python
+num_epochs = 2
+num_batches = 10
+data_loader = DataLoader(num_batches)
+communicator = AllreduceCommunicator()
+
+for _ in range(num_epochs):
+    for features, labels in data_loader:
+        outputs = model(features)
+        loss = loss_fn(outputs, labels)
+        grads = calculate_gradients(loss, model)
```

Comment:
Should we calculate the gradients of all the parameters of the model. And then we can execute AllReduce?

Reply:
Yes this should be a blocking call until all gradients are calculated in the workers.